src/libopensc/pkcs15-cac.c

```diff
@@ -0,0 +1,429 @@
+#define MANU_ID		"Common Access Card"
+
+int sc_pkcs15emu_cac_init_ex(sc_pkcs15_card_t *, sc_pkcs15emu_opt_t *);
+
+
+
+typedef struct pdata_st {
+	const char *id;
+	const char *label;
+	const char *path;
+	int         ref;
+	int         type;
+	unsigned int maxlen;
+	unsigned int minlen;
+	unsigned int storedlen;
+	int         flags;	
+	int         tries_left;
+	const unsigned char  pad_char;
+	int         obj_flags;
+} pindata; 
+
+static int cac_detect_card(sc_pkcs15_card_t *p15card)
+{
+	sc_card_t *card = p15card->card;
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+	if (card->type < SC_CARD_TYPE_CAC_GENERIC
+		|| card->type >= SC_CARD_TYPE_CAC_GENERIC+1000)
+		return SC_ERROR_INVALID_CARD;
+	return SC_SUCCESS;
+}
+
+#define CAC_NUM_CERTS_AND_KEYS 10
+
+const char * cac_get_name(int type)
+{
+    switch (type) {
+    case SC_CARD_TYPE_CAC_I: return ("CAC I");
+    case SC_CARD_TYPE_CAC_II: return ("CAC II");
+    default: break;
+    }
+    return ("CAC");
+}
+
+/*
+ * These could move to a helper file for other cards that wish to use usage as a way of getting flags
+ */
+
+/* Only certain usages are valid for a given algorithm, return all the usages that the algorithm supports so we
+ * can use it as a filter for all the public and private key usages */
+static unsigned int
+cac_alg_flags_from_algorithm(int algorithm)
+{
+	switch (algorithm) {
+#ifdef SC_ALGORITHM_RSA
+	case SC_ALGORITHM_RSA:
+		return SC_PKCS15_PRKEY_USAGE_ENCRYPT | SC_PKCS15_PRKEY_USAGE_WRAP |
+		       SC_PKCS15_PRKEY_USAGE_VERIFY | SC_PKCS15_PRKEY_USAGE_VERIFYRECOVER |
+		       SC_PKCS15_PRKEY_USAGE_DECRYPT | SC_PKCS15_PRKEY_USAGE_UNWRAP |
+		       SC_PKCS15_PRKEY_USAGE_SIGN | SC_PKCS15_PRKEY_USAGE_SIGNRECOVER |
+		       SC_PKCS15_PRKEY_USAGE_NONREPUDIATION;
+#endif
+#ifdef SC_ALGORITHM_DSA
+	case SC_ALGORITHM_DSA:
+		return SC_PKCS15_PRKEY_USAGE_VERIFY| SC_PKCS15_PRKEY_USAGE_SIGN | 
+		       SC_PKCS15_PRKEY_USAGE_NONREPUDIATION;
+#endif
+#ifdef SC_ALGORITHM_DH
+	case SC_ALGORITHM_DH:
+		return SC_PKCS15_PRKEY_USAGE_DERIVE ;
+#endif
+#ifdef SC_ALGORITHM_EC
+	case SC_ALGORITHM_EC:
+		return SC_PKCS15_PRKEY_USAGE_DERIVE | SC_PKCS15_PRKEY_USAGE_VERIFY|
+		       SC_PKCS15_PRKEY_USAGE_SIGN | SC_PKCS15_PRKEY_USAGE_NONREPUDIATION;
+#endif
+#ifdef SC_ALGORITHM_GOSTR3410
+	case SC_ALGORITHM_GOSTR3410:
+		return SC_PKCS15_PRKEY_USAGE_DERIVE | SC_PKCS15_PRKEY_USAGE_VERIFY|
+		       SC_PKCS15_PRKEY_USAGE_SIGN | SC_PKCS15_PRKEY_USAGE_NONREPUDIATION;
+#endif
+	}
+	return 0;
+}
+
+
+/* These are the cert key usage bits that map to various PKCS #11 (and thus PKCS #15) flags */
+#define SC_CERT_USAGE_SIGNATURE 0x63ULL /* Signature, Non repudication, Cert Sign, Crl sign */
+#define SC_CERT_USAGE_DERIVE    0x10ULL /* Key agreement */
+#define SC_CERT_USAGE_UNWRAP	0x14ULL /* Key encipherment, Key agreement */
+#define SC_CERT_USAGE_DECRYPT	0x88ULL /* Data encipherment, encipher */
+#define SC_CERT_USAGE_NONREPUDIATION 0x02ULL
```

Comment:
@dengert The PIV driver have the usage flags set based on the ID and on a hard-coded list in `pkcs15-piv.c` (knows all its possible objects that can be on the card and their attributes). There is no such table for CAC and the usage flags are derived from the X509 key usage flags.

This is what is one of the differences between PIV and CAC. It can be theoretically used in other cards, if we will move these constants and helper functions out of `pkcs15-cac.c`, but it would need somebody with a bit better overview about the whole project.